### PR TITLE
symengine 0.8.1 (new formula)

### PR DESCRIPTION
**Pull request by ProfFan**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

### CODE REVIEW

File: Formula/symengine.rb [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+  test do
+    (testpath/"test.cpp").write <<~EOS
+      #include <symengine/expression.h>
+      using SymEngine::Expression;
+      int main() {
+        Expression x("x");
+        auto ex = pow(x+sqrt(Expression(2)), 6);
+      }
+    EOS
+    gmp_flags = shell_output("pkg-config --libs gmp").split
```
Comment (carlocab) [inline Formula/symengine.rb:26]: You don't have access to `pkg-config` in the test, so this isn't going to work (as you can see from the CI failure).

```suggestion
    gmp_flags = ["-L#{Formula["gmp"].opt_lib", "-lgmp"]
```

**Review by carlocab — COMMENTED** [review]
Thanks for your PR. I have a few small suggestions. When you make further changes, please use the guidelines in our commit style guide referenced in the PR template.

Here's a quick summary:

> Please use the preferred commit-message style for homebrew/core. We put the name of the formula first in commit-message headings.
> 
> For new formulae:
> 
> > At Homebrew, we like to put the name of the formula up front like so: `foobar 7.3 (new formula)`.
> 
> For existing formulae:
> 
> > The preferred commit message format for simple version updates is `foobar 7.3` and for fixes is `foobar: fix flibble matrix.`.
> 
> Refer to the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit) for more details. Also, when making further changes to your pull request, use the following guidelines to make sure that @BrewTestBot can merge your commits:
> 
> > - One formula per commit; one commit per formula.
> > - Keep merge commits out of the pull request.

File: Formula/symengine.rb [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+    system "cmake", "--install", "build"
+  end
+
+  test do
+    (testpath/"test.cpp").write <<~EOS
+      #include <symengine/expression.h>
+      using SymEngine::Expression;
+      int main() {
+        Expression x("x");
+        auto ex = pow(x+sqrt(Expression(2)), 6);
```
Comment (carlocab) [inline Formula/symengine.rb:23]: Can we rewrite this to produce output we can check? Or at least, make this `return 1` if the contents of `ex` aren't what's expected.

File: Formula/symengine.rb (hunk 1 above)
Comment (ProfFan) [reply Formula/symengine.rb:26]: Thanks! I think this is something should be added to the new formula guide and Formula cookbook (I searched for how to do this for ages)

File: Formula/symengine.rb [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+      }
+    EOS
+    gmp_flags = ["-L#{Formula["gmp"].opt_lib", "-lgmp"]
```
Comment (carlocab) [inline Formula/symengine.rb:26]:
```suggestion
    gmp_flags = ["-L#{Formula["gmp"].opt_lib}", "-lgmp"]
```

File: Formula/symengine.rb (hunk 2 above)
Comment (ProfFan) [reply Formula/symengine.rb]: done

File: Formula/symengine.rb [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+  depends_on "cmake" => :build
+  depends_on "flint"
+  depends_on "gmp"
+  depends_on "mpfr"
+
+  def install
+    system "cmake", "-S", ".", "-B", "build", \
+    "-DBUILD_SHARED_LIBS=ON", "-DWITH_GMP=ON", "-DWITH_MPFR=ON", \
+    "-DWITH_MPC=ON", "-DINTEGER_CLASS=flint", "-DWITH_SYMENGINE_THREAD_SAFE=ON", \
+    *std_cmake_args
```
Comment (carlocab) [inline Formula/symengine.rb:17]:
```suggestion
    system "cmake", "-S", ".", "-B", "build",
                    "-DBUILD_SHARED_LIBS=ON",
                    "-DWITH_GMP=ON",
                    "-DWITH_MPFR=ON",
                    "-DWITH_MPC=ON",
                    "-DINTEGER_CLASS=flint",
                    "-DWITH_SYMENGINE_THREAD_SAFE=ON",
                    *std_cmake_args
```

File: Formula/symengine.rb [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+        auto ex = x+sqrt(Expression(2))+1;
+        auto equality = eq(ex+1, expand(ex));
+        return equality == true;
+      }
+    EOS
+    gmp_flags = ["-L#{Formula["gmp"].opt_lib}", "-lgmp"]
+    mpfr_flags = ["-L#{Formula["mpfr"].opt_lib}", "-lmpfr"]
+    flint_flags = ["-L#{Formula["flint"].opt_lib}", "-lflint"]
+    system ENV.cc, "test.cpp", "-std=c++11", "-L#{lib}", "-lsymengine", \
+    "-lc++", *flint_flags, *mpfr_flags, *gmp_flags, "-o", "test"
```
Comment (carlocab) [inline Formula/symengine.rb:38]:
```suggestion
    lib_flags = [
      "-L#{Formula["gmp"].opt_lib}", "-lgmp",
      "-L#{Formula["mpfr"].opt_lib}", "-lmpfr",
      "-L#{Formula["flint"].opt_lib}", "-lflint"
    ]
    system ENV.cxx, "test.cpp", "-std=c++11", "-L#{lib}", "-lsymengine", *lib_flags, "-o", "test"
```

**Review by carlocab — COMMENTED** [review]
You're probably missing a
```ruby
depends_on "libmpc"
```

**Review by carlocab — COMMENTED** [review]
Please use the preferred commit-message style for homebrew/core. We put the name of the formula first in commit-message headings.

For new formulae:

> At Homebrew, we like to put the name of the formula up front like so: `foobar 7.3 (new formula)`.

For existing formulae:

> The preferred commit message format for simple version updates is `foobar 7.3` and for fixes is `foobar: fix flibble matrix.`.

Refer to the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit) for more details. Also, when making further changes to your pull request, use the following guidelines to make sure that @BrewTestBot can merge your commits:

> - One formula per commit; one commit per formula.
> - Keep merge commits out of the pull request.

**Review by carlocab — COMMENTED** [review]
Thanks, @ProfFan. One last question, while we're waiting for CI to complete. Upstream recommends using LLVM: https://github.com/symengine/symengine#for-package-managers

Why aren't we doing so here?

**Comment by ProfFan** [comment]
@carlocab Upstream support for LLVM is lagging so not possible to compile with current LLVM 13. I can try fixing that for upstream but for now not happening.

**Comment by ProfFan** [comment]
Or* I can add a dependency for `llvm@12`. What do you think @carlocab ?

**Comment by carlocab** [comment]
I see. Is adding LLVM 12 as a dependency useful? What features does it enable?

**Comment by ProfFan** [comment]
OK, I have figured it out. It's an upstream issue that got fixed in master but not 0.8.1 release.

I added the hot fix for it according to upstream issues.

File: Formula/symengine.rb [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+  def install
+    system "cmake", "-S", ".", "-B", "build",
+                    "-DBUILD_SHARED_LIBS=ON",
+                    "-DWITH_GMP=ON",
+                    "-DWITH_MPFR=ON",
+                    "-DWITH_MPC=ON",
+                    "-DINTEGER_CLASS=flint",
+                    "-DWITH_LLVM=ON",
+                    "-DWITH_COTIRE=OFF",
+                    "-DLLVM_DIR=#{Formula["llvm"].opt_lib}/cmake/llvm",
```
Comment (carlocab) [inline Formula/symengine.rb:23]: This needs to be `llvm@12` if you need LLVM 12. You're also missing a
```ruby
depends_on "llvm@12`
```

Comment (ProfFan) [reply Formula/symengine.rb:24]: Sorry, I mean with the fix it builds with LLVM 13, I also added the dependency.

File: Formula/symengine.rb [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+  license "MIT"
+
+  depends_on "cmake" => :build
+  depends_on "flint"
+  depends_on "gmp"
+  depends_on "libmpc"
+  depends_on "llvm"
+  depends_on "mpfr"
+
+  def install
```
Comment (carlocab) [inline Formula/symengine.rb:15]:
```suggestion
  depends_on "mpfr"

  fails_with gcc: "5"

  def install
```
This should help with the Linux build failure.

**Review by carlocab — APPROVED** [review]
Thanks for your work here, and for your first contribution to Homebrew!

**Comment by ProfFan** [comment]
Thanks for the suggestions @carlocab !

**Pull request closed without merging**